### PR TITLE
add arcface

### PR DESCRIPTION
**Pull request by okotaku**
## Motivation

Other implementation of arcface. I suggest this style.
Related to https://github.com/open-mmlab/mmclassification/pull/926 .

mmengine hasn't released yet, so I haven't checked tests pass.

## Use cases (Optional)

I used this implement at [kaggle happy whale competition](https://www.kaggle.com/competitions/happy-whale-and-dolphin) .

```
head=dict(
                 type='NormLinearClsHead',
                 num_classes=15587,
                 in_channels=5504,
                 feature_norm=True,
                 weight_norm=True,
                 bias=False,
                 loss=dict(type='ArcFaceLoss'),
                 topk=(1, 5),
                 cal_acc=True
             )
```

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmcls/models/heads/norm_linear_head.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        return cls_score
+
+    def predict(
+            self,
+            feats: Tuple[torch.Tensor],
+            data_samples: List[ClsDataSample] = None) -> List[ClsDataSample]:
+        """Test without augmentation."""
+        # The part can be traced by torch.fx
+        cls_score = self(feats)
+        cls_score = cls_score * self.compute_loss.s
```
Comment (Ezra-Yu) [inline mmcls/models/heads/norm_linear_head.py:83]: Thanks for your contribution.

Arcface as I understand it has two parts, one is the Norm part, and one is the projection function part like `softmax`. The loss used is still CE.

In your implementation, there is no way for the output to pass through the projection function. I checked the [relevant implementation](https://github.com/deepinsight/insightface/blob/babb9a58bbc42ae4b648acdbb803159a35f53db3/recognition/arcface_torch/losses.py#L89), and the output needs to go through the projection function. Therefore, I think Arcface is not suitable as a loss.

However, your NormClsHead is a good design, we are considering it, but it also has shortcomings, 1. It will be difficult for users to find ArcFace unless there are many similar ArcFace structures that use NormClsHead; 2. When testing, the weight will calculate the norm once in each forward.

Comment (okotaku) [reply mmcls/models/heads/norm_linear_head.py:83]: Arcface flow is following.

Training,
norm weights & norm logits -> cosine -> add margin based on class label -> scale -> loss

Test,
norm weights & norm logits -> cosine -> scale

In my implementation, `NormLinearClsHead` covers `norm weights & norm logits -> cosine` for training and `norm weights & norm logits -> cosine -> scale` for testing. And `ArcFaceLoss` covers `cosine -> add margin based on class label -> scale -> loss`.

Comment (okotaku) [reply mmcls/models/heads/norm_linear_head.py:83]: Do you think `Head` should have `norm weights & norm logits -> cosine -> add margin based on class label -> scale` too, and `ArcFaceHead / CosFaceHead / ArcFaceAdaptiveMarginHead / etc... + CELoss` is better?

Comment (Ezra-Yu) [reply mmcls/models/heads/norm_linear_head.py:83]: > Arcface flow is following.
> 
> Training, norm weights & norm logits -> cosine -> add margin based on class label -> scale -> loss
> 
> Test, norm weights & norm logits -> cosine -> scale

Thank you very much for explaining the ArcFace detailedly. Sorry that there are some mistakes in my previous understanding.

> Do you think `Head` should have `norm weights & norm logits -> cosine -> add margin based on class label -> scale` too, and `ArcFaceHead / CosFaceHead / ArcFaceAdaptiveMarginHead / etc... + CELoss` is better?

Yes, I prefer `ArcFaceHead / CosFaceHead / ArcFaceAdaptiveMarginHead / etc... + CELoss` , I think [`self.compute_loss.s`](https://github.com/RistoranteRist/mmclassification/blob/34d3bee1009feb8e0716d8738c0f6eda44f2c2f1/mmcls/models/heads/norm_linear_head.py#L83) is unexpected and users may use ArcFace in other head such as the `linear-head`. And I look up other impls, such as [insight-face](https://github.com/deepinsight/insightface/blob/babb9a58bbc42ae4b648acdbb803159a35f53db3/recognition/arcface_torch/partial_fc_v2.py#L165), [adaFace](https://github.com/mk-minchul/AdaFace), are all separated from loss module.
